File: src/main/java/com/qtech/bubbles/event/Bus.java
package com.qtech.bubbles.event;

import com.qtech.bubbles.QBubbles;
import com.qtech.bubbles.addon.loader.AddonContainer;
import com.qtech.bubbles.common.addon.AddonObject;
import com.qtech.bubbles.common.addon.QBubblesAddon;
import com.qtech.bubbles.event.bus.AddonEventBus;
import com.qtech.bubbles.event.bus.LocalAddonEventBus;
import com.qtech.bubbles.event.bus.QBubblesEventBus;
import org.jetbrains.annotations.NotNull;

@SuppressWarnings("unused")
public class Bus {
    private static final AddonEventBus addonEventBus = new AddonEventBus();

    public static LocalAddonEventBus<?> getLocalAddonEventBus(String addonId) {
        AddonContainer container = QBubbles.getInstance().getAddonManager().getContainerFromId(addonId);
        AddonObject<? extends QBubblesAddon> addonObject;
        if (container != null) {
            addonObject = container.getAddonObject();
            return addonObject.getEventBus();
        }
        throw new NullPointerException();
    }

    public static @NotNull LocalAddonEventBus<? extends QBubblesAddon> getLocalAddonEventBus(QBubblesAddon addon) {
        AddonObject<? extends QBubblesAddon> addonObject = addon.getAddonObject();
        return addonObject.getEventBus();
    }

    public static <T extends QBubblesAddon> LocalAddonEventBus<T> getLocalAddonEventBus(AddonObject<T> addonObject) {
        return addonObject.getEventBus();
    }

    public static AddonEventBus getAddonEventBus() {
        return addonEventBus;
    }

    public static @NotNull QBubblesEventBus getQBubblesEventBus() {
        return QBubbles.getEventBus();
    }
}

File: src/main/kotlin/com/qtech/bubbles/event/Bus.kt
package com.qtech.bubbles.event

import com.qtech.bubbles.BubbleBlaster
import com.qtech.bubbles.addon.loader.AddonContainer
import com.qtech.bubbles.common.addon.AddonObject
import com.qtech.bubbles.common.addon.AddonInstance
import com.qtech.bubbles.event.bus.AddonEventBus
import com.qtech.bubbles.event.bus.BBEventBus
import com.qtech.bubbles.event.bus.LocalAddonEventBus

object Bus {
    @kotlin.jvm.JvmStatic
    val addonEventBus = AddonEventBus()
    fun getLocalAddonEventBus(addonId: String): LocalAddonEventBus<*> {
        val container: AddonContainer? = BubbleBlaster.instance.addonManager.getContainer(addonId)
        val addonInstanceObject: AddonObject<out AddonInstance> = container!!.obj
        return addonInstanceObject.eventBus
    }

    fun getLocalAddonEventBus(addonInstance: AddonInstance): LocalAddonEventBus<out AddonInstance> {
        val addonObject = addonInstance.`object`
        return addonObject.eventBus
    }

    fun <T : AddonInstance> getLocalAddonEventBus(addonObject: AddonObject<T>): LocalAddonEventBus<T> {
        return addonObject.eventBus
    }

    @kotlin.jvm.JvmStatic
    val qBubblesEventBus: BBEventBus
        get() = BubbleBlaster.eventBus
}